### PR TITLE
bareos-fd-postgres: added support for PostgreSQL 15 and later

### PR DESCRIPTION
**Pull request by lluuaapp**
After upgrading to Debian bookworm I realized that the bareos-fd-postgres plugin is no longer compatible with version 15 of PostgreSQL as it

1. no longer creates the `backup_label` file
2. renames `pg_start_backup` to `pg_backup_start`
3. renames `pg_stop_backup` to `pg_backup_stop`

See https://www.postgresql.org/docs/current/release-15.html (look for the non-exclusive backup mode feature).

This PR ignores the `backup_label` file for PostgreSQL 15 and later and uses the result of the `pg_backup_stop` SELECT statement. It also uses the new function names when starting/stoping the backup.

**Backport of PR #0000 to bareos-2x** (remove this line if this is no backport; for backport use cherry-pick -x)

#### Please check

- [x] Short description and the purpose of this PR is present _above this paragraph_
- [ ] Your name is present in the AUTHORS file (optional)

If you have any questions or problems, please give a comment in the PR.

### Helpful documentation and best practices

- [Git Workflow](https://docs.bareos.org/DeveloperGuide/gitworkflow.html)
- [Automatic Sourcecode Formatting](https://docs.bareos.org/DeveloperGuide/generaldevel.html#automatic-sourcecode-formatting)
- [Check your commit messages](https://docs.bareos.org/DeveloperGuide/gitworkflow.html#commits)
- [Boy Scout Rule](https://docs.bareos.org/DeveloperGuide/generaldevel.html#boy-scout-rule)

### Checklist for the _reviewer_ of the PR (will be processed by the Bareos team)
Make sure you check/merge the PR using `devtools/pr-tool` to have some simple automated checks run and a proper changelog record added.

##### General
- [x] Is the PR title usable as CHANGELOG entry?
- [x] Purpose of the PR is understood
- [x] Commit descriptions are understandable and well formatted
- [x] Check backport line
- [ ] Required backport PRs have been created

##### Source code quality
- [ ] Source code changes are understandable
- [ ] Variable and function names are meaningful
- [ ] Code comments are correct (logically and spelling)
- [ ] Required documentation changes are present and part of the PR

##### Tests
- [ ] Decision taken that a test is required (if not, then remove this paragraph)
- [ ] The choice of the type of test (unit test or systemtest) is reasonable
- [ ] Testname matches exactly what is being tested
- [ ] On a fail, output of the test leads quickly to the origin of the fault

### CODE REVIEW

**Comment by lluuaapp** [comment]
Sorry, seems that I do not have the right so view the details of the failed checks. Either you have to invite me to seem them or add them here as a comment. When reviewing my changes I can not find anything that should not run in Python 2.7...

Thanks!

**Comment by bruno-at-bareos** [comment]
@lluuaapp it will be picked automatically by our CI Jenkins. 
Let see how it goes.

**Comment by bruno-at-bareos** [comment]
@lluuaapp While Jenkins has its own life, I will have a few remarks about code and commit message reveal by our tools (see devtools/pr-tool) and documentation https://docs.bareos.org/bareos-22/DeveloperGuide/gitworkflow.html#commits especially length.

```
 ✗  Commit checks failed:
        a5ef440c7 bareos-fd-postgres: hopefully fixed build on EL_7 with python2.7: headline too long
        9a70a4b27 bareos-fd-postgres: using result of pg_backup_stop instead of backup_label for postgres 15 and later: headline too long
        e59bbeb04 bareos-fd-postgres: fixed pg_backup_start/pg_start_backup and pg_backup_stop/pg_stop_backup for postgres 15: headline too long
 ✗  bareos-check-sources --since=7b84f946cb895d7bf758e45f15130a23d7e16926 reported:
        Plugin 'fix copyright year' would modify 'core/src/plugins/filed/python/postgres/BareosFdPluginPostgres.py'
        Plugin 'python black' would modify 'core/src/plugins/filed/python/postgres/BareosFdPluginPostgres.py'
        Plugin 'fix copyright year' would modify 'systemtests/tests/py3plug-fd-postgres/database/setup_local_db.sh.in'

**Comment by lluuaapp** [comment]
Sorry, no, I did not run the system tests nor did I test a full restore. Do not have the capabilities here, running on a production system.

**Comment by bruno-at-bareos** [comment]
@lluuaapp thanks for the feedback this help to understand what's going on.

In the pg_backup_stop stage it look like you didn't follow exactly the recommendation
`The pg_stop_backup will return one row with three values. The second of these fields should be written to a file named backup_label in the root directory of the backup. The third field should be written to a file named tablespace_map unless the field is empty. These files are vital to the backup working and must be written byte for byte without modification, which may require opening the file in binary mode.`

As such the backup_label file is missing in restore, and as such the restored cluster can't start until you recreate that file.
(there a .backup file into wal archives that also can help to recover manually).

I believe that def parseBackupStopResult needs to be redesigned and improved.
I will already push some minor fixes

**Comment by bruno-at-bareos** [comment]
Here the first bunch of patches, (I don't have the right to push to your branch)
[patches.zip](https://github.com/bareos/bareos/files/11801783/patches.zip)

**Comment by lluuaapp** [comment]
> As such the backup_label file is missing in restore, and as such the restored cluster can't start until you recreate that file.
(there a .backup file into wal archives that also can help to recover manually).

Maybe you are right and the fix was/is much more complicated than I thought.

Should I apply the patches and update the PR or do you want me to close this PR and you can create a new one with all changes?

**Comment by bruno-at-bareos** [comment]
@lluuaapp hello Paul, I've retried the push and this time made it successful, so the patches are history.

So yes the fix is a bit more complicated as it seems, but still certainly worse the effort.
We will just find a way to store the data available at pg_backup_stop and to reuse them on restore stage.

**Comment by bruno-at-bareos** [comment]
@lluuaapp would you mind to finalize the missing part: mean implement https://stackoverflow.com/a/75076640 advise ?

**Comment by lluuaapp** [comment]
@bruno-at-bareos I will try to fix this this week. As I am on vacation for a few weeks from next week on it will last a bit longer to properly fix this.

**Comment by lluuaapp** [comment]
@bruno-at-bareos as you may have seen I put some work into this. I have some questions though:

1. Location of the backup_label/tablespace_map files:

From the PostgreSQL documentation:

> The second of these fields should be written to a file named backup_label in the root directory of the backup. The third field should be written to a file named tablespace_map unless the field is empty. These files are vital to the backup working and must be written byte for byte without modification, which may require opening the file in binary mode.

I am not sure whether the "root directory of the backup" is the same as the data directory for bareos. Also (see 2.) it would make sense to create the files somewhere else in the filesystem but restore them into the proper folder(s) later.

2. Secure creation/deletion of the auxiliary files

We must make absolutely sure to delete the files (backup_label, tablespace_map and recovery.signal). If they exist and the DB process is restarted (e.g. for an upgrade etc) it will not startup properly. Maybe you have improvements for this?

I could not find a way to `self.files_to_backup.append()` in a way that the file has a different location on disk than in the backup itself. This maybe the safest way, though.

3. Is there a reason why we require pg8000 v1.16 and later? Debian Bookworm comes with v1.10.6 and you have to `pip3 install --break-system-packages pg8000` to get the proper version (and `break-system-packages` is something most user should not like to do).

4. Again, as I do not have a test system I could not test a full cycle (backup and restore). Did I get this correctly that this is something the unit tests do?

Thanks,
Paul

**Comment by bruno-at-bareos** [comment]
Hi @lluuaapp Thanks for pushing this forward.

From easier to harder answer. Yes the systemtests to a full run of the test so create a cluster backup it, destroy it, restore, start the cluster again, check it is up and data are there.

Normally PG15 will take care of the recovery.signal and remove alone the backup_label. I wouldn't take care too much about this, at least from a Bareos point of view, once everything that is needed to start the pg cluster in recovery mode is present on disk, the job is done ;-)

for pg8000 there's function used not present in 1.10 (actually pg8000 is well maintained and actual version is 1.29) I would ping debian maintainers why they don't update pg8000 package, not the other way around ;-)

For the backup_label and table space, I have one of my colleague that will help me and you to understand the concept of virtual file to be backup, and then can be restored. But unfortunately he's on holiday for a still a few weeks. 
So I guess we will wait until his and your return, except if I get information how to do this.

File: core/src/plugins/filed/python/postgres/BareosFdPluginPostgres.py [hunk 1]
```diff
@@ -456,38 +494,86 @@ def close_db_connection(self):
...
+            stopStmt = "pg_backup_stop(wait_for_archive => true)"
+        else:
+            stopStmt = "pg_stop_backup()"
+            self.parseBackupLabelFile()
+
         try:
-            results = self.dbCon.run("SELECT pg_stop_backup();")
-            self.lastLSN = self.formatLSN(results[0][0])
+            results = self.dbCon.run("SELECT %s;" % stopStmt)
+            firstRow = results[0][0]
```
Comment (bruno-at-bareos) [inline core/src/plugins/filed/python/postgres/BareosFdPluginPostgres.py:548]: shouldn't first row being results[0] ? 
results[0][0] being the first column of the first row.

**Comment by lluuaapp** [comment]
Hi @bruno-at-bareos,

sorry, could not find your comment in the "Files changed" section to answer directly. Seem I have either not the rights or Github is buggy?

Anyway, I know its strange, but when I tested the code

`firstRow = results[0][0]`
 
with only one `[0]` it crashes (something like "expected string not tuple"). It makes no sense to me, too but `bCon.run()` returns a 3D array in this case.

I have another question. Did setup a test system now and would like to run the tests. Can you tell me how to enable the `py3plug-fd-postgres` test?

`ctest --show-only`

does not show the test (neither enabled nor disabled).

Thanks,
Paul

**Comment by bruno-at-bareos** [comment]
> Hi @bruno-at-bareos,
> 
> sorry, could not find your comment in the "Files changed" section to answer directly. Seem I have either not the rights or Github is buggy?
> 
> Anyway, I know its strange, but when I tested the code
> 
> `firstRow = results[0][0]`
> 
> with only one `[0]` it crashes (something like "expected string not tuple"). It makes no sense to me, too but `bCon.run()` returns a 3D array in this case.
> 
> I have another question. Did setup a test system now and would like to run the tests. Can you tell me how to enable the `py3plug-fd-postgres` test?
> 
> `ctest --show-only`
> 
> does not show the test (neither enabled nor disabled).
> 
> Thanks, Paul

A quick reply the test will be enable automatically if the python3 dev stack is there and also mandatory package
pg8000 and python-dateutil

**Comment by bruno-at-bareos** [comment]
Hi @lluuaapp 

It look like I was able to finish your changes, here at least locally I was able to run the full tests, and backup and restore where working. What we missed was the fact that backup_label file has to be written only when the Full (pg base_backup) is done.
Not during incremental where only the Wal are in used.

Maybe you will be able to reproduce the success, which would be a nice improvement.

To speed up tests, I've added the fast => True on pg_backup_start function so it create the checkpoint as soon as the backup start. That need to be moved to option so people can choose whatever is the most convenient for us.

Our CI is having some troubles actually, I hope we will fix it soon.

**Comment by bruno-at-bareos** [comment]
Hello Paull @lluuaapp

We have raised the concern internally about the support of old exclusive mode, and newver non exclusive mode.
Trying to do both in the same plugin will be really complicated, especially in terms of restore.

We decided to go for a new plugin which should be called postgresql which will do only non exclusive backup supporting all PostgreSQL  >= 10 version.

I will transfer the new work into a new branch and PR today or monday.
Then I will close this PR. 

We would like to thank you a lot for starting contributing, and hope to have you joining the beta testing of the new plugin once ready.

Regards.

**Comment by lluuaapp** [comment]
Hi @bruno-at-bareos,

that is perfectly fine with me. Also thought that these if/else and PG version checks makes the code even more complicated.

Looking forward to test the beta version. Just let me know when it is ready (or where I have to subscribe to get notified).

Thanks,
Paul

**Comment by bruno-at-bareos** [comment]
As discussed previously, closing the PR.
See you soon in the next plugin.

**Pull request closed without merging**